src/csharp/Grpc.Examples/MathGrpc.cs

```diff
@@ -27,10 +27,73 @@ public static partial class Math
   {
     static readonly string __ServiceName = "math.Math";
 
+    #if !GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION
+    static readonly grpc::Marshaller<global::Math.DivArgs> __Marshaller_math_DivArgs = 
```

Comment:
Here's an example of the generated code. Note that serialization seems broken at the moment.

Reply:
The buffer marshaller methods are quite long. @jtattermusch what do you think of including a private method here that the marshallers call so their delegates are only 1 line? Generated code will be slightly longer if there is only one message type, but as soon as there are multiple then the reusable method will greatly reduce amount of code gen.

```cs
private static void WriteBufferMessage<T>(T message, SerializationContext context) where T : IBufferMessage
{
    var writer = new global::Google.Protobuf.CodedOutputWriter(context.GetBufferWriter());
    arg.WriteTo(ref writer);
    writer.Flush();
    context.Complete();
}
```

Don't need one for reading. Code generation can use parser's `IMessageParser<T>.ParseFrom(ReadOnlySequence<byte>)`:

```cs
(context) => global::Math.DivArgs.Parser.ParseFrom(context.PayloadAsReadOnlySequence())
```

Before:

```csharp
    static readonly grpc::Marshaller<global::Math.DivReply> __Marshaller_math_DivReply = 
      grpc::Marshallers.Create(
        (arg, context) =>
        {
          var writer = new global::Google.Protobuf.CodedOutputWriter(context.GetBufferWriter());
          arg.WriteTo(ref writer);
          writer.Flush();
          context.Complete();
        },
        context =>
        {
          var result = new global::Math.DivReply();
          var reader = new global::Google.Protobuf.CodedInputReader(context.PayloadAsReadOnlySequence());
          result.MergeFrom(ref reader);
          return result;
        });
```

After:

```csharp
    static readonly grpc::Marshaller<global::Math.DivReply> __Marshaller_math_DivReply = grpc::Marshallers.Create(WriteBufferMessage, (context) => global::Math.DivReply.Parser.ParseFrom(context.PayloadAsReadOnlySequence());
```

Reply:
For parsing, you can just use 
`global::Math.DivReply.Parser.ParseFrom(ctx.PayloadAsReadOnlySequence());` (very similar pattern as before, using the Parser static property of the message).   The `IMessageParser<T>.ParseFrom(ReadOnlySequence<byte>)` proposed by @JamesNK might work as well, but using the Parser property seems cleaner.

Reply:
If we are generating a helper method then I think we could add fallback logic for non-IBufferMessage types.

```csharp
private static void WriteBufferMessage(T message, SerializationContext context) where T : IMessage
{
    if (message is IBufferMessage bufferMessage)
    {
        var writer = new global::Google.Protobuf.CodedOutputWriter(context.GetBufferWriter());
        arg.WriteTo(ref writer);
        writer.Flush();
        context.Complete();
    }
    else
    {
        context.Complete(global::Google.Protobuf.MessageExtensions.ToByteArray(message));
    }
}
```

What version of C# do we target in code generation? This `(message is IBufferMessage bufferMessage)` is relatively modern.

Reply:
On avoiding conflicts, the static type has:
* `__ServiceName`
* `__Marshaller_*`
* `__Method_*`
* `Descriptor`
* `SERVICENAMEBase`
* `SERVICENAMEClient`

If we name this method `__Helper_WriteBufferMessage` then we won't conflict with anything.